src/main/java/com/google/cloud/genomics/dataflow/readers/bam/ReadBAMTransform.java

```diff
@@ -0,0 +1,142 @@
+import com.google.cloud.genomics.utils.Contig;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Takes a tuple of 2 collections: Contigs and BAM files and transforms them into
+ * a collection of reads by reading BAM files in a sharded manner.
+ */
+public class ReadBAMTransform extends PTransform<PCollectionTuple, PCollection<Read>> {
+  GenomicsFactory.OfflineAuth auth;
+  
+  public static TupleTag<Contig> CONTIGS_TAG = new TupleTag<>();
+  public static TupleTag<String> BAMS_TAG = new TupleTag<>();
+  
+  public static PCollection<Read> getReadsFromBAMFilesSharded(
+      Pipeline p, 
+      GenomicsFactory.OfflineAuth auth,
+      Iterable<Contig> contigs, List<String> BAMFiles) {
+      ReadBAMTransform readBAMSTransform = new ReadBAMTransform();
+      readBAMSTransform.setAuth(auth);
+      PCollectionTuple tuple = PCollectionTuple
+          .of(
+              ReadBAMTransform.BAMS_TAG, 
+                p.apply(
+                    Create.of(BAMFiles))
+                .setCoder(StringUtf8Coder.of()))
+         .and(ReadBAMTransform.CONTIGS_TAG, 
+             p.apply(
+                 Create.of(
+                     contigs))
+             .setCoder(SerializableCoder.of(Contig.class)));
+      return readBAMSTransform.apply(tuple);
+  }
+  
+  public static class ShardFn extends DoFn<String, BAMShard> {
+    GenomicsFactory.OfflineAuth auth;
+    PCollectionView<Iterable<Contig>, ?> contigsView;
+    Storage.Objects storage;
+    
+    public ShardFn(GenomicsFactory.OfflineAuth auth, PCollectionView<Iterable<Contig>, ?> contigsView) {
+      this.auth = auth;
+      this.contigsView = contigsView;
+    }
+    
+    @Override
+    public void startBundle(DoFn<String, BAMShard>.Context c) throws IOException {
+      storage = GCSOptions.Methods.createStorageClient(c, auth);
+    }
+    
+    @Override
+    public void processElement(ProcessContext c) throws java.lang.Exception {
+      (new Sharder(storage, c.element(), c.sideInput(contigsView), c))
+          .process();
+    }
+  }
+  
+  public static class ReadFn extends DoFn<BAMShard, Read> {
+    GenomicsFactory.OfflineAuth auth;
+    Storage.Objects storage;
+    
+    public ReadFn(GenomicsFactory.OfflineAuth auth) {
+      this.auth = auth;
+
+    }
+    
+    @Override
+    public void startBundle(DoFn<BAMShard, Read>.Context c) throws IOException {
+      storage = GCSOptions.Methods.createStorageClient(c, auth);
+    }
+    
+    @Override
+    public void processElement(ProcessContext c) throws java.lang.Exception {
+      (new Reader(storage, c.element(), c))
+          .process();
+    }
+  }
+  
+  
+  @Override
+  public PCollection<Read> apply(PCollectionTuple contigsAndBAMs) {
+
+    final PCollection<Contig> contigs = contigsAndBAMs.get(CONTIGS_TAG);
+    final PCollectionView<Iterable<Contig>, ?> contigsView =
```

Comment:
Why are contigsView being passed as a pcollection tuple here, but as a side input to the next transform?  I think a side input would have been clearer here.